# THỰC HÀNH HỒI QUY TUYẾN TÍNH ĐA BIẾN

Sử dụng bộ dữ liệu Boston trong gói lệnh MASS bao gồm 14 biến liên quan đến giá trị nhà ở vùng ngoại ô ở Boston và hàm step, phân tích hồi quy bội của biến medv (giá nhà trung bình – đơn vị: nghìn $) theo các biến còn lại.

In [1]:
# install.packages("MASS")
library(MASS)

a) Đưa ra mô hình hồi quy tuyến tính "forward" và "backward" tốt nhất.

In [2]:
only = lm(medv ~ 1, data = Boston)
all = lm(medv ~ ., data = Boston)

In [3]:
forward = step(only, formula(all), direction = "forward", trace = 0)
forward$coefficients

(Intercept)         lstat            rm       ptratio           dis 
 36.341145004  -0.522553457   3.801578840  -0.946524570  -1.492711460 
          nox          chas         black            zn          crim 
-17.376023429   2.718716303   0.009290845   0.045844929  -0.108413345 
          rad           tax 
  0.299608454  -0.011777973

In [4]:
backward = step(all, formula(only), direction = "backward", trace = 0)
backward$coefficients

(Intercept)          crim            zn          chas           nox 
 36.341145004  -0.108413345   0.045844929   2.718716303 -17.376023429 
           rm           dis           rad           tax       ptratio 
  3.801578840  -1.492711460   0.299608454  -0.011777973  -0.946524570 
        black         lstat 
  0.009290845  -0.522553457

b) Khi phân tích “forward”, nếu biến medv được biểu diễn theo hai biến thì đó
là những biến nào?

In [5]:
forward$anova

Step      Df Deviance    Resid. Df Resid. Dev AIC     
1            NA          NA 505       42716.30   2246.514
2  + lstat   -1 23243.91400 504       19472.38   1851.009
3  + rm      -1  4033.07222 503       15439.31   1735.577
4  + ptratio -1  1711.32389 502       13727.99   1678.131
5  + dis     -1   499.07761 501       13228.91   1661.393
6  + nox     -1   759.56355 500       12469.34   1633.473
7  + chas    -1   328.27141 499       12141.07   1621.973
8  + black   -1   272.83713 498       11868.24   1612.473
9  + zn      -1   189.93614 497       11678.30   1606.309
10 + crim    -1    94.71193 496       11583.59   1604.189
11 + rad     -1   228.60431 495       11354.98   1596.103
12 + tax     -1   273.61928 494       11081.36   1585.761

Hai biến đó là lstat và rm

c) Khi phân tích “backward”, kiểm định xem các hệ số trong mô hình hồi quy tuyến tính thu được có thực sự khác 0 không? Phần dư có tuân theo phân phối chuẩn với giá trị trung bình bằng 0 không?

In [6]:
summary(backward)


Call:
lm(formula = medv ~ crim + zn + chas + nox + rm + dis + rad + 
    tax + ptratio + black + lstat, data = Boston)

Residuals:
     Min       1Q   Median       3Q      Max 
-15.5984  -2.7386  -0.5046   1.7273  26.2373 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)  36.341145   5.067492   7.171 2.73e-12 ***
crim         -0.108413   0.032779  -3.307 0.001010 ** 
zn            0.045845   0.013523   3.390 0.000754 ***
chas          2.718716   0.854240   3.183 0.001551 ** 
nox         -17.376023   3.535243  -4.915 1.21e-06 ***
rm            3.801579   0.406316   9.356  < 2e-16 ***
dis          -1.492711   0.185731  -8.037 6.84e-15 ***
rad           0.299608   0.063402   4.726 3.00e-06 ***
tax          -0.011778   0.003372  -3.493 0.000521 ***
ptratio      -0.946525   0.129066  -7.334 9.24e-13 ***
black         0.009291   0.002674   3.475 0.000557 ***
lstat        -0.522553   0.047424 -11.019  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*

H0: Hệ số bằng 0  
H1: Hệ số khác 0  
Do các hệ số đều có p-value < 0,05 $\Rightarrow$ Bác bỏ H0  
Có cơ sở để nói các hệ số trong mô hình "backward" là khác 0

In [7]:
shapiro.test(backward$residuals)


	Shapiro-Wilk normality test

data:  backward$residuals
W = 0.90131, p-value < 2.2e-16


H0: Phần dư tuân theo phân phối chuẩn  
H1: Phần dư không tuân theo phân phối chuẩn  
Do p-value < 2.2e-16 < 0.05 $\Rightarrow$ Bác bỏ H0  
Có cơ sở để nói phần dư của mô hình không tuân theo phân phối chuẩn

In [8]:
wilcox.test(backward$residuals)


	Wilcoxon signed rank test with continuity correction

data:  backward$residuals
V = 55447, p-value = 0.008285
alternative hypothesis: true location is not equal to 0


H0: Phần dư có trung bình bằng 0  
H1: Phần dư có trung bình khác 0  
Do p-value = 0.008285 < 0.05 $\Rightarrow$ Bác bỏ H0  
Có cơ sở để nói phần dư của mô hình là khác 0